Bug: testutil package returning unexpected results

<!--

	For bug reports, please provide the version or the Git commit hash for
	which you have observed the behavior in question.

	Note that we use GitHub issues for bugs and (uncontroversial) feature
	requests (see below for details).

	Please do *NOT* ask usage questions in GitHub issues. Usage questions make
	most sense on the users mailing list, where more people are available to
	potentially respond to your question, and the whole community can benefit
	from the answers provided (perhaps your question has already been answered,
	search the archive to find out):
	https://groups.google.com/forum/#!forum/prometheus-users

	While a GitHub issue is fine to track progress on an uncontroversial
	feature request, many feature requests touch the best practices and
	concepts of Prometheus as a whole and need to be discussed with the wider
	developer community first. This is in particular true for a request to
	reconsider a prior rejection of a feature request. Those overarching
	discussions happen on the developer mailing list (GitHub issues, in
	particular closed ones, are not tracked by the wider developer community
	and thus inadequate):
	https://groups.google.com/forum/#!forum/prometheus-developers
	
	You can find more information at: https://prometheus.io/community/

-->

I'm new to working with Prometheus so apologies if this is a misunderstanding on my part. 

I noticed when writing a test for a Counter without any ConstLabels defined in the CounterOpts struct, the test will fail and provide an error message saying two identical strings do not match.

This can be replicated by modifying slightly one of the tests in the `testutil_test.go` file, by just removing the labels, example:

```go
func TestCollectAndCompareBug(t *testing.T) {
	const metadata = `
		# HELP some_total A value that represents a counter.
		# TYPE some_total counter
	`

	c := prometheus.NewCounter(prometheus.CounterOpts{
		Name: "some_total",
		Help: "A value that represents a counter.",
	})
	c.Inc()

	expected := `

		some_total 1
	`

	if err := CollectAndCompare(c, strings.NewReader(metadata+expected), "some_total"); err != nil {
		t.Errorf("unexpected collecting result:\n%s", err)
	}
}
```

This results in the following output:

```bash

--- FAIL: TestCollectAndCompareBug (0.00s)
    testutil_test.go:165: unexpected collecting result:
        
        metric output does not match expectation; want:
        
        # HELP some_total A value that represents a counter.
        # TYPE some_total counter
        some_total 1
        
        
        got:
        
        # HELP some_total A value that represents a counter.
        # TYPE some_total counter
        some_total 1
        
FAIL

Process finished with exit code 1
```

I haven't dug in too deeply but if this is a bug and anyone have some pointers to fix I can take it on as soon as I have some free cycles.

--- a/prometheus/testutil/testutil.go
+++ b/prometheus/testutil/testutil.go
@@ -37,7 +37,6 @@ import (
 	"bytes"
 	"fmt"
 	"io"
-	"reflect"
 
 	"github.com/prometheus/common/expfmt"
 
@@ -125,39 +124,49 @@ func CollectAndCompare(c prometheus.Collector, expected io.Reader, metricNames .
 // exposition format. If any metricNames are provided, only metrics with those
 // names are compared.
 func GatherAndCompare(g prometheus.Gatherer, expected io.Reader, metricNames ...string) error {
-	metrics, err := g.Gather()
+	got, err := g.Gather()
 	if err != nil {
 		return fmt.Errorf("gathering metrics failed: %s", err)
 	}
 	if metricNames != nil {
-		metrics = filterMetrics(metrics, metricNames)
+		got = filterMetrics(got, metricNames)
 	}
 	var tp expfmt.TextParser
-	expectedMetrics, err := tp.TextToMetricFamilies(expected)
+	wantRaw, err := tp.TextToMetricFamilies(expected)
 	if err != nil {
 		return fmt.Errorf("parsing expected metrics failed: %s", err)
 	}
+	want := internal.NormalizeMetricFamilies(wantRaw)
 
-	if !reflect.DeepEqual(metrics, internal.NormalizeMetricFamilies(expectedMetrics)) {
-		// Encode the gathered output to the readable text format for comparison.
-		var buf1 bytes.Buffer
-		enc := expfmt.NewEncoder(&buf1, expfmt.FmtText)
-		for _, mf := range metrics {
-			if err := enc.Encode(mf); err != nil {
-				return fmt.Errorf("encoding result failed: %s", err)
-			}
+	if len(got) != len(want) {
+		return notMatchingError(got, want)
+	}
+	for i := range got {
+		if got[i].String() != want[i].String() {
+			return notMatchingError(got, want)
 		}
-		// Encode normalized expected metrics again to generate them in the same ordering
-		// the registry does to spot differences more easily.
-		var buf2 bytes.Buffer
-		enc = expfmt.NewEncoder(&buf2, expfmt.FmtText)
-		for _, mf := range internal.NormalizeMetricFamilies(expectedMetrics) {
-			if err := enc.Encode(mf); err != nil {
-				return fmt.Errorf("encoding result failed: %s", err)
-			}
-		}
+	}
+	return nil
+}
 
-		return fmt.Errorf(`
+// notMatchingError encodes both provided slices of metric families into the
+// text format and creates a readable error message from the result.
+func notMatchingError(got, want []*dto.MetricFamily) error {
+	var gotBuf, wantBuf bytes.Buffer
+	enc := expfmt.NewEncoder(&gotBuf, expfmt.FmtText)
+	for _, mf := range got {
+		if err := enc.Encode(mf); err != nil {
+			return fmt.Errorf("encoding gathered metrics failed: %s", err)
+		}
+	}
+	enc = expfmt.NewEncoder(&wantBuf, expfmt.FmtText)
+	for _, mf := range want {
+		if err := enc.Encode(mf); err != nil {
+			return fmt.Errorf("encoding expected metrics failed: %s", err)
+		}
+	}
+
+	return fmt.Errorf(`
 metric output does not match expectation; want:
 
 %s
@@ -165,9 +174,7 @@ metric output does not match expectation; want:
 got:
 
 %s
-`, buf2.String(), buf1.String())
-	}
-	return nil
+`, wantBuf.String(), gotBuf.String())
 }
 
 func filterMetrics(metrics []*dto.MetricFamily, names []string) []*dto.MetricFamily {

--- a/prometheus/testutil/testutil_test.go
+++ b/prometheus/testutil/testutil_test.go
@@ -143,6 +143,28 @@ func TestCollectAndCompare(t *testing.T) {
 	}
 }
 
+func TestCollectAndCompareNoLabel(t *testing.T) {
+	const metadata = `
+		# HELP some_total A value that represents a counter.
+		# TYPE some_total counter
+	`
+
+	c := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "some_total",
+		Help: "A value that represents a counter.",
+	})
+	c.Inc()
+
+	expected := `
+
+		some_total 1
+	`
+
+	if err := CollectAndCompare(c, strings.NewReader(metadata+expected), "some_total"); err != nil {
+		t.Errorf("unexpected collecting result:\n%s", err)
+	}
+}
+
 func TestNoMetricFilter(t *testing.T) {
 	const metadata = `
 		# HELP some_total A value that represents a counter.

